src/core/ccitt_stream.js

```diff
@@ -0,0 +1,60 @@
+/* Copyright 2012 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Dict, isDict } from './primitives';
+import { CCITTFaxDecoder } from './ccitt';
+import { DecodeStream } from './stream';
+
+var CCITTFaxStream = (function CCITTFaxStreamClosure() {
+  function CCITTFaxStream(str, maybeLength, params) {
+    this.str = str;
+    this.dict = str.dict;
+
+    if (!isDict(params)) {
+      params = Dict.empty;
+    }
+    this.ccittFaxDecoder = new CCITTFaxDecoder(str, {
+      K: params.get('K'),
+      EndOfLine: params.get('EndOfLine'),
+      EncodedByteAlign: params.get('EncodedByteAlign'),
+      Columns: params.get('Columns'),
+      Rows: params.get('Rows'),
+      EndOfBlock: params.get('EndOfBlock'),
+      BlackIs1: params.get('BlackIs1'),
+    });
+
+    DecodeStream.call(this, maybeLength);
+  }
+
+  CCITTFaxStream.prototype = Object.create(DecodeStream.prototype);
+
+  CCITTFaxStream.prototype.readBlock = function() {
+    while (!this.eof) {
+      let c = this.ccittFaxDecoder.readNextChar();
+      if (c === -1) {
```

Comment:
This "if" is not present in the original code. Do you know what is changed?

Reply:
I think that the old code was actually wrong here, since `lookChar` used to return `null` when EOF was reached. Hence we would actually attempt to insert `null` into `this.buffer` here, since the `!this.eof` check didn't occur until the *next* loop iteration.

In the new code, we first of all need to set `this.eof = true;` in the `CCITTFaxStream`, since various stream methods expect that property to exist. Secondly, with the refactoring `readNextChar` will return `-1` in the EOF case, and we *really* don't want insert that into `this.buffer`. (Considering that setting a `uint8array[i] = -1` will lead to `uint8array[i} === 255`.)